### ISSUE TITLE
Sampling rate below 32768Hz plays some instruments out of tune with certain SoundFonts

### ISSUE THREAD

**Issue by fdelapena** [FluidSynth/fluidsynth#1059]
### FluidSynth version
```
FluidSynth runtime version 2.2.5
Copyright (C) 2000-2022 Peter Hanappe and others.
Distributed under the LGPL license.
SoundFont(R) is a registered trademark of Creative Technology Ltd.

FluidSynth executable version 2.2.5
Sample type=double
```

### Describe the bug
For some reason, some notes play sound some semitones different for certain instruments when setting specific sampling rate with a particular midi file and SoundFont.

### Expected behavior
FluidSynth shouldn't play some detuned instruments when changing sampling rate with certain SoundFonts and files.

### Steps to reproduce
Unzip the following attachments:
[issue.zip](https://github.com/FluidSynth/fluidsynth/files/8165366/issue.zip)
[soundfont.zip](https://github.com/FluidSynth/fluidsynth/files/8165386/soundfont.zip)

Run:
```
fluidsynth -C0 -R0 -r32767 soundfont.sf2 issue.mid
```
After some seconds when the lead chord enters, note samples sound clearly out of tune.

### Additional context
If the sampling rate value is `32768` Hz or above, the issue does not get reproduced.

**Comment by derselbst** [comment]
Interesting. I can confirm the issue. It happens at 14 seconds when the third MIDI channel starts playing. Program 84 is playing. The corresponding instrument in the soundfont specifies the "modEnvToPitch" initial generator to be 500 cents, i.e. +5 semitones. 

I'll still have to figure out why  this generator may or may not be respected depending on the sample rate.

**Comment by mawe42** [comment]
I have the feeling the problem could be in [`calculate_hold_decay_buffers`](https://github.com/FluidSynth/fluidsynth/blob/3ac9b6fbf054c9cd3b8facad92e4993731625073/src/synth/fluid_voice.c#L725):
```
/* round to next full number of buffers */
buffers = (int)(((fluid_real_t)voice->output_rate * seconds)
                        / (fluid_real_t)FLUID_BUFSIZE
                        + 0.5f);
```

The comment above that line seems to suggest that the aim was to *round up* to the nearest full buffer count (which seems to make sense, as the aim is to calculate buffer length?). But the code actually rounds fractions < `.5` down. And some of those envelopes have `timecents` of `-12000.f`, which is `0.000977` in seconds.

`0.000977 * 32768 = 0.5 => 1 buffer`
`0.000977 * 32767 = 0.499985 => 0 buffers`

If I add a `ceil` before the cast to `int` and remove the `+ 0.5f`, the reported problem goes away. But I'm not sure what other repercussions this change might have...

**Comment by derselbst** [comment]
Closely connected to your finding Marcus, I think I found a nicer fix:

https://github.com/FluidSynth/fluidsynth/blob/3ac9b6fbf054c9cd3b8facad92e4993731625073/src/rvoice/fluid_adsr_env.h#L77-L82

@diwic Actually was wondering already 12 years ago, why this is only applied to volenv. When also applying it to modenv, the envelope is correctly progressing from attack to decay and finally stays in sustain. It sounds fine to me in this case.

I still need to comprehend, if there's a possibility for the envelope getting stuck in the attack phase though...

**Comment by derselbst** [comment]
I'm confused. In case of attack and release phases fluidsynth is unconditionally applying [+1 to the sample increment](https://github.com/FluidSynth/fluidsynth/blob/3ac9b6fbf054c9cd3b8facad92e4993731625073/src/synth/fluid_voice.c#L1081) to avoid a division by zero which in turn also prevents the envelope from being stuck in any of those two phases. But why isn't that done for decay as well? Instead, zero sample increment is explicitly permitted by `count ? -1.0f / count : 0.0f`:

https://github.com/FluidSynth/fluidsynth/blob/3ac9b6fbf054c9cd3b8facad92e4993731625073/src/synth/fluid_voice.c#L1100

Perhaps the +1 increment isn't done, because the code used for decay is also used by two other switch-case labels? I understand that zero-increment is permitted for sustain, because that's just an amplitude.

And because of this "not applying +1" (or ceiling as you suggested Marcus) in case of decay, we then have to do the
`"If we're switching envelope stages from decay to sustain, force the value to be the end value of the previous stage"`-dance above.

Btw, those lines were presumably introduced by Christian's legendary EMU10K1 patch:
https://github.com/FluidSynth/fluidsynth/commit/f89377de888dc21e5364cf990d90e4af01c2e3fd#diff-350915195f3eae7c0981c83bfbd76f6df5da1c2726890ad85b4f49420395dfeaR286-R289

@mrbumpy409 In case you still remember: Can you confirm that the motivation behind introducing those three lines linked in your patch, were to prevent the decay phase from being stuck?

**Comment by jjceresa** [comment]
>I still need to comprehend, if there's a possibility for the envelope getting stuck in the attack phase though...

Yes there is a possibility at the end of HOLD to get the envelope stuck to 1.0f. This occurs when DECAY `count` is 0;  in this case `inc` is 0.0f and the sustain level will never be reached. This is why the `"switching envelope stages from decay to sustain, force the value..."` dance is required.  And I see no reason not to apply this `dance`  to modenv.

**Comment by mawe42** [comment]
Implementing the workaround for stuck envelope phases to modenv as well sounds good. Well, not "good" as in "the right solution". But probably the easiest and least intrusive solution.

I still think it's a little weird that we effectively use `1 + floor((output_rate * seconds) / bufsize)` for some phases and `round((output_rate * seconds) / bufsize)` for others, then special case the `0` result.

**Comment by derselbst** [comment]
> I still think it's a little weird that we effectively use `1 + floor((output_rate * seconds) / bufsize)` for some phases and `round((output_rate * seconds) / bufsize)` for others, then special case the `0` result.

I feel the same way. But in the end I've also preferred the least intrusive change here: #1067

**Comment by jjceresa** [comment]
Of course Implementing the `workaround for stuck envelope` isn't the perfect solution. Even this solution could produce bad effect in the case of low Sustain level. For example if the `sustain level is 25%`, and `decay count is 0` then at the end of `hold` segment the envelope will suddenly fall from 1.0f to 0.25f with a potential glitch.
